server/services/ai.ts

```diff
@@ -1,33 +1,185 @@
+// Use the writeToDB parameters to skip (or introduce them where needed).
+// Otherwise, changes will get duplicated.
+export default class AIService {
     starUpgradeService: StarUpgradeService;
+    carrierService: CarrierService;
+    starService: StarService;
+    distanceService: DistanceService;
+    waypointService: WaypointService;
+    combatService: CombatService;
+    shipTransferService: ShipTransferService;
+    technologyService: TechnologyService;
+    playerService: PlayerService;
+    reputationService: ReputationService;
+    diplomacyService: DiplomacyService;
+    playerStatisticsService: PlayerStatisticsService;
 
     constructor(
-        starUpgradeService: StarUpgradeService
+        starUpgradeService: StarUpgradeService,
+        carrierService: CarrierService,
+        starService: StarService,
+        distanceService: DistanceService,
+        waypointService: WaypointService,
+        combatService: CombatService,
+        shipTransferService: ShipTransferService,
+        technologyService: TechnologyService,
+        playerService: PlayerService,
+        reputationService: ReputationService,
+        diplomacyService: DiplomacyService,
+        playerStatisticsService: PlayerStatisticsService
     ) {
         this.starUpgradeService = starUpgradeService;
+        this.carrierService = carrierService;
+        this.starService = starService;
+        this.distanceService = distanceService;
+        this.waypointService = waypointService;
+        this.combatService = combatService;
+        this.shipTransferService = shipTransferService;
+        this.technologyService = technologyService;
+        this.playerService = playerService;
+        this.reputationService = reputationService;
+        this.diplomacyService = diplomacyService;
+        this.playerStatisticsService = playerStatisticsService;
     }
 
     async play(game: Game, player: Player) {
         if (!player.defeated) {
             throw new Error('The player is not under AI control.');
         }
 
-        let isFirstTick = game.state.tick % game.settings.galaxy.productionTicks === 1;
-        let isLastTick = game.state.tick % game.settings.galaxy.productionTicks === game.settings.galaxy.productionTicks - 1;
+        const isFirstTick = game.state.tick % game.settings.galaxy.productionTicks === 1;
+        const isLastTick = game.state.tick % game.settings.galaxy.productionTicks === game.settings.galaxy.productionTicks - 1;
 
-        if (isFirstTick) {
-            await this._playFirstTick(game, player);
-        } else if (isLastTick) {
-            await this._playLastTick(game, player);
+        if (game.settings.general.advancedAI === 'enabled') {
+            await this._doAdvancedLogic(game, player, isFirstTick, isLastTick);
+        }
+
+        await this._doBasicLogic(game, player, isFirstTick, isLastTick);
+    }
+
+    async _doBasicLogic(game: Game, player: Player, isFirstTick: boolean, isLastTick: boolean) {
+        try {
```

Comment:
This try/catch stuff would be better placed in the `play` function above and wrap both `doAdvancedLogic` and `doBasicLogic`.